clients/client-py/taskcluster/aio/upload.py

```diff
@@ -0,0 +1,133 @@
+"""
+Support for uploading objects to the object service, following best
+practices for that service.
+
+Data for upload is read from a "reader" provided by a "reader factory".  A
+reader has an async `read(max_size)` method which reads and returns a chunk of
+1 .. `max_size` bytes, or returns an empty string at EOF.  A reader factory is an async
+callable which returns a fresh reader, ready to read the first byte of the
+object.  When uploads are retried, the reader factory may be called more than
+once.
+
+Note that `aiofile.open` returns a value suitable for use as a reader, if async
+file IO is important to the application.
+
+This module provides several pre-defined readers and reader factories for
+common cases.
+"""
+import six
+
+if six.PY2:
+    raise ImportError("upload is only supported in Python 3")
+
+import base64
+
+import aiohttp
+
+import taskcluster
+from .reader_writer import streamingCopy, BufferReader, BufferWriter, FileReader
+from .retry import retry
+
+DATA_INLINE_MAX_SIZE = 8192
+
+
+async def upload_from_buf(*, data, **kwargs):
+    """
+    Convenience method to upload data from an in-memory buffer.  Arguments are the same
+    as `upload` except that `readerFactory` should not be supplied.
+    """
+    async def readerFactory():
+        return BufferReader(data)
+
+    await upload(**kwargs, readerFactory=readerFactory)
+
+
+async def upload_from_file(*, file, **kwargs):
+    """
+    Convenience method to upload data from a a file.  The file should be open
+    for reading, in binary mode, and be seekable (`f.seek`).  Remaining
+    arguments are the same as `upload` except that `readerFactory` should not
+    be supplied.
+    """
+    async def readerFactory():
+        file.seek(0)
+        return FileReader(file)
+
+    await upload(**kwargs, readerFactory=readerFactory)
+
+
+async def upload(*, projectId, name, contentType, contentLength, expires,
+                 readerFactory, maxRetries=5, objectService):
+    """
+    Upload the given data to the object service with the given metadata.
+    The `maxRetries` parameter has the same meaning as for service clients.
+    The `objectService` paraemter is an instance of the Object class,
+    configured with credentials for the upload.
+    """
+    async with aiohttp.ClientSession() as session:
+        uploadId = taskcluster.slugid.v4()
```

Comment:
Maybe more future-proof to use `taskcluster.slugid.nice()` here? At the moment it probably isn't needed, but 121 bits entropy should still be plenty, and if a command line tool in future passes an `uploadId` as its first argument at some point in future, it has a little extra protection.